<a href="https://colab.research.google.com/github/Gizeh-Gutierrez/Maestria_programacion/blob/main/S06_Actividad_02_ALGEBRALINEAL_271752.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Actividad calificada — Álgebra lineal con NumPy

**Programación para Analítica Descriptiva y Predictiva**
Maestría en Inteligencia Artificial y Analítica de Datos — UACJ

---

**Material de estudio previo:** `Material_Extra_Algebra_Lineal_NumPy.ipynb`. Revísalo antes de resolver esta actividad.

**Entrega:** individual, en este notebook de Google Colab (comparte el enlace con permiso de edición para el docente).
**Plazo:** 7 días naturales a partir de la publicación de esta actividad.
**Penalización por entrega tardía:** 5% por día, hasta 7 días; después de eso, la actividad no se recibe.
**Valor total:** 100 puntos.

### Herramientas permitidas

Esta actividad evalúa `np.linalg.norm()`, `np.dot()`/`@`, `np.linalg.det()`, `np.linalg.inv()`, `np.linalg.solve()`, y las operaciones de NumPy ya vistas en sesiones anteriores. No uses pandas ni scikit-learn.

### Cómo se evalúa

Cada ejercicio indica su valor en puntos. Se evalúa el resultado correcto **y** el uso de la función específica de `np.linalg` indicada en cada enunciado — calcular a mano una fórmula que NumPy ya resuelve (por ejemplo, programar la fórmula del determinante 2×2 en vez de usar `np.linalg.det()`) no obtiene el puntaje completo.

In [2]:
import numpy as np

## Parte 1 — Vectores: perfiles de producto

Una tienda en línea representa cada producto como un vector de 3 componentes: calificación de calidad, calificación de precio, y calificación de servicio (escala 1 a 5, donde 5 es mejor). Ejecuta la siguiente celda para cargar los tres productos — **no la modifiques**.

In [3]:
producto_A = np.array([4.5, 3.2, 5.0])
producto_B = np.array([4.0, 2.8, 4.7])
producto_C = np.array([1.0, 5.0, 1.2])

print(producto_A, producto_B, producto_C)

[4.5 3.2 5. ] [4.  2.8 4.7] [1.  5.  1.2]


---

### Ejercicio 1 — Norma de los vectores (10 pts)

Calcula la norma de `producto_A`, `producto_B` y `producto_C` usando `np.linalg.norm()`.

In [4]:
# Tu código aquí

norma_A = np.linalg.norm(producto_A)
norma_B = np.linalg.norm(producto_B)
norma_C = np.linalg.norm(producto_C)

print("Norma de producto_A:", norma_A)
print("Norma de producto_B:", norma_B)
print("Norma de producto_C:", norma_C)

Norma de producto_A: 7.449161026585477
Norma de producto_B: 6.777167549942971
Norma de producto_C: 5.238320341483518


---

### Ejercicio 2 — Producto punto y ángulo (15 pts)

1. Calcula el producto punto entre `producto_A` y `producto_B`, y entre `producto_A` y `producto_C`, usando `np.dot()` o `@`.
2. Calcula el ángulo (en grados) entre `producto_A` y `producto_B`, y entre `producto_A` y `producto_C`.
3. Con base en los ángulos obtenidos, indica (en un comentario o celda de texto) cuál de los dos productos (`B` o `C`) es más similar a `producto_A`, y por qué el ángulo es el indicador correcto para esa comparación.

In [7]:
# Tu código aquí
# Producto punto
AB = np.dot(producto_A, producto_B)
AC = np.dot(producto_A, producto_C)

print("Producto punto AB =", AB)
print("Producto punto AC =", AC)

# Angulos
angulo_AB = np.degrees(np.arccos(AB / (norma_A * norma_B)))
angulo_AC = np.degrees(np.arccos(AC / (norma_A * norma_C)))

print("angulo entre AB =", angulo_AB)
print("angulo entre AC =", angulo_AC)

Producto punto AB = 50.46
Producto punto AC = 26.5
angulo entre AB = 1.7745815839558179
angulo entre AC = 47.22509268261831


El producto B es mas similar al producto A, porque el angulo entre sus vectores es menor 1.77 que el ángulo entre A y C 47.23

---

### Ejercicio 3 — Normalización (10 pts)

Normaliza los tres vectores (`producto_A`, `producto_B`, `producto_C`) para obtener sus vectores unitarios. Verifica, imprimiendo el resultado, que la norma de cada vector normalizado sea 1.

In [8]:
# Tu código aquí
# Normalizacion
A_unitario = producto_A / np.linalg.norm(producto_A)
B_unitario = producto_B / np.linalg.norm(producto_B)
C_unitario = producto_C / np.linalg.norm(producto_C)

# vectores unitarios
print("A unitario:", A_unitario)
print("B unitario:", B_unitario)
print("C unitario:", C_unitario)

# verificar que sea 1
print("Norma de A unitario:", np.linalg.norm(A_unitario))
print("Norma de B unitario:", np.linalg.norm(B_unitario))
print("Norma de C unitario:", np.linalg.norm(C_unitario))


A unitario: [0.60409488 0.42957858 0.67121653]
B unitario: [0.59021707 0.41315195 0.69350506]
C unitario: [0.19090089 0.95450444 0.22908106]
Norma de A unitario: 1.0
Norma de B unitario: 1.0
Norma de C unitario: 1.0


## Parte 2 — Matrices: mezcla de insumos

Una fábrica de jugos combina dos insumos (concentrado de fruta y agua) en distintas proporciones para dos líneas de producto. La matriz `M` representa esas proporciones (filas = línea de producto, columnas = insumo), y `P` representa un ajuste de escala para una nueva presentación. Ejecuta la siguiente celda — **no la modifiques**.

In [9]:
M = np.array([[2, 1],
              [1, 3]])

P = np.array([[1.5, 0],
              [0, 2.0]])

print(M)
print(P)

[[2 1]
 [1 3]]
[[1.5 0. ]
 [0.  2. ]]


---

### Ejercicio 4 — Multiplicación matricial (15 pts)

1. Calcula `M * P` (elemento a elemento) y `M @ P` (multiplicación matricial). Los resultados son distintos — en un comentario o celda de texto, explica brevemente por qué.
2. Verifica que las formas sean compatibles: ¿cuál sería el resultado de `P @ M`? Calcúlalo y compáralo con `M @ P` — ¿da el mismo resultado?

In [10]:
# Tu código aquí

elemento = M * P


MP = M @ P
PM = P @ M

print("M * P :")
print(elemento)
print("M @ P:")
print(MP)
print("P @ M:")
print(PM)

M * P :
[[3. 0.]
 [0. 6.]]
M @ P:
[[3.  2. ]
 [1.5 6. ]]
P @ M:
[[3.  1.5]
 [2.  6. ]]


La diferencia se debe a que M*P multiplica posición por posición, mientras que M@P multiplica matriz por matriz es decir fila × columna


P @ M es diferente de M @ P
porque la multiplicación de matrices no es conmutativa.

---

### Ejercicio 5 — Determinante e inversa (15 pts)

1. Calcula el determinante de `M` con `np.linalg.det()`. Confirma que `M` sí tiene inversa (determinante distinto de 0).
2. Calcula la inversa de `M` con `np.linalg.inv()`.
3. Verifica tu resultado calculando `M @ M_inv` — debe dar (aproximadamente) la matriz identidad.

In [12]:
# Tu código aquí
det_M = np.linalg.det(M)
print("determinante de M:", det_M)

# inversa
M_inv = np.linalg.inv(M)
print("inversa de M:")
print(M_inv)

# Verificar
identidad = M @ M_inv
print("M @ Minv:")
print(identidad)

determinante de M: 5.000000000000001
inversa de M:
[[ 0.6 -0.2]
 [-0.2  0.4]]
M @ Minv:
[[ 1.00000000e+00  0.00000000e+00]
 [-5.55111512e-17  1.00000000e+00]]


---

### Ejercicio 6 — Sistema de ecuaciones (15 pts)

La fábrica necesita producir un lote que requiere en total 180 unidades de la línea 1 y 300 unidades de la línea 2, según las proporciones de `M`. Esto se representa como el sistema $Mx = b$, con:

```python
b = np.array([180, 300])
```

1. Resuelve el sistema con `np.linalg.solve()` para encontrar cuántas unidades de cada insumo (`x`) se necesitan.
2. Verifica tu solución multiplicando `M @ x` y confirmando que el resultado coincide con `b`.

In [14]:
b = np.array([180, 300])

# Tu código aquí
x = np.linalg.solve(M, b)
print(" x:")
print(x)

#verificar
verificacion = M @ x
print("M @ x =")
print(verificacion)
print("coincide con b?", np.allclose(verificacion, b))

 x:
[48. 84.]
M @ x =
[180. 300.]
coincide con b? True


---

### Ejercicio 7 — Síntesis: distancia y similitud (20 pts)

Retomando los tres productos de la Parte 1:

1. Calcula la **distancia euclidiana** entre `producto_A` y `producto_B`, y entre `producto_A` y `producto_C`, usando la norma de la diferencia entre los vectores (`np.linalg.norm(a - b)`).
2. Calcula la **similitud coseno** entre los mismos pares — es el cociente que ya calculaste en el Ejercicio 2 antes de aplicar `arccos()`: $\frac{a \cdot b}{\|a\| \|b\|}$.
3. Con base en ambas métricas (distancia y similitud coseno), concluye en una cadena de texto impresa cuál producto (`B` o `C`) es más parecido a `producto_A`, y si ambas métricas coinciden en la misma conclusión.

In [18]:
# Tu código aquí
# distancia euclidiana
dist_AB = np.linalg.norm(producto_A - producto_B)
dist_AC = np.linalg.norm(producto_A - producto_C)

# similitud coseno
cos_AB = np.dot(producto_A, producto_B) / (
np.linalg.norm(producto_A) * np.linalg.norm(producto_B)
)

cos_AC = np.dot(producto_A, producto_C) / (
np.linalg.norm(producto_A) * np.linalg.norm(producto_C)
)
print("distancia AB:", dist_AB)
print("distancia AC:", dist_AC)
print("similitud coseno AB:", cos_AB)
print("similitud coseno AC:", cos_AC)



cercano_distancia = "B" if dist_AB < dist_AC else "C"
cercano_coseno = "B" if cos_AB > cos_AC else "C"

if cercano_distancia == cercano_coseno:
  print(
  f"el producto {cercano_distancia} es el mas parecido a producto A "
  "ambas métricas coinciden"
)
else:
  print(
      f"la distancia euclidiana muestra que el producto {mas_cercano_distancia} "
      f"es mas parecido a producto A, mientras que la similitud coseno indica "
      f"que es el producto {mas_cercano_coseno}. Las metricas no coinciden."
)

distancia AB: 0.7071067811865476
distancia AC: 5.470831746635972
similitud coseno AB: 0.9995203969214979
similitud coseno AC: 0.6791199021706347
el producto B es el mas parecido a producto A ambas métricas coinciden


---

## Antes de entregar

- Verifica que el notebook completo corra de principio a fin sin errores (**Entorno de ejecución → Ejecutar todas**).
- Confirma que no usaste pandas ni scikit-learn.
- Comparte el notebook con permiso de **edición** para el docente.